File: frontend/src/types/resolve.ts
export interface ProductCandidate {
  product_id: string;
  title: string;
  score: number;
  confidence_band: string;
  rationale: string;
  matched_aliases: string[];
  rules_fired: string[];
}

export interface VariantCandidate {
  variant_id: string;
  product_id: string;
  sku: string;
  color?: string | null;
  size?: string | null;
  normalized_color?: string | null;
  normalized_size?: string | null;
  available_stock: number;
  score: number;
  confidence_band: string;
  rationale: string;
  matched_aliases: string[];
  rules_fired: string[];
}

export interface ResolveProductResponse {
  trace_id: string;
  query: string;
  candidates: ProductCandidate[];
  confidence_band: string;
  confidence_score: number;
  missing_slots: string[];
  rationale?: string | null;
}

export interface ResolveVariantResponse {
  trace_id: string;
  product_id?: string | null;
  candidates: VariantCandidate[];
  confidence_band: string;
  confidence_score: number;
  missing_slots: string[];
  rationale?: string | null;
}

export interface ResolverTrace {
  id: string;
  shop_id: string;
  trace_type: string;
  conversation_id?: string | null;
  input_payload: Record<string, unknown>;
  top_candidates: Record<string, unknown>[];
  matched_aliases: Record<string, unknown>[];
  rules_fired: string[];
  missing_slots: string[];
  confidence_band: string;
  confidence_score: number;
  rationale?: string | null;
  qdrant_query_metadata: Record<string, unknown>;
  created_at: string;
}

export interface ResolverFeedbackRequest {
  shop_id: string;
  action: 'accept_ai' | 'correct_product' | 'correct_variant' | 'taxonomy_issue';
  original_product_id?: string;
  corrected_product_id?: string;
  original_variant_id?: string;
  corrected_variant_id?: string;
  notes?: string;
}

export interface ResolverFeedback {
  id: string;
  shop_id: string;
  trace_id: string;
  action: string;
  operator_id: string;
  notes?: string | null;
  created_at: string;
}

File: backend/app/schemas/resolve.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class ResolveMediaReference(BaseModel):
    media_id: str | None = None
    media_url: str | None = None


class ResolveConversationContext(BaseModel):
    conversation_id: UUID | None = None
    prior_product_id: UUID | None = None
    prior_messages: list[str] = Field(default_factory=list)
    extracted_slots: dict = Field(default_factory=dict)


class ResolveProductRequest(BaseModel):
    shop_id: UUID
    message_text: str
    media_references: list[ResolveMediaReference] = Field(default_factory=list)
    conversation_context: ResolveConversationContext | None = None
    limit: int = Field(default=5, ge=1, le=20)
    fusion_strategy: str | None = None


class ProductCandidate(BaseModel):
    product_id: UUID
    title: str
    score: float
    confidence_band: str
    rationale: str
    matched_aliases: list[str] = Field(default_factory=list)
    rules_fired: list[str] = Field(default_factory=list)


class ResolveProductResponse(BaseModel):
    trace_id: UUID
    query: str
    candidates: list[ProductCandidate]
    confidence_band: str
    confidence_score: float
    missing_slots: list[str] = Field(default_factory=list)
    rationale: str | None = None


class ResolveVariantRequest(BaseModel):
    shop_id: UUID
    message_text: str
    product_id: UUID | None = None
    candidate_product_ids: list[UUID] = Field(default_factory=list)
    media_references: list[ResolveMediaReference] = Field(default_factory=list)
    conversation_context: ResolveConversationContext | None = None
    raw_color: str | None = None
    raw_size: str | None = None
    quantity: int = Field(default=1, ge=1)
    limit: int = Field(default=5, ge=1, le=20)


class VariantCandidate(BaseModel):
    variant_id: UUID
    product_id: UUID
    sku: str
    color: str | None = None
    size: str | None = None
    normalized_color: str | None = None
    normalized_size: str | None = None
    available_stock: int
    score: float
    confidence_band: str
    rationale: str
    matched_aliases: list[str] = Field(default_factory=list)
    rules_fired: list[str] = Field(default_factory=list)


class ResolveVariantResponse(BaseModel):
    trace_id: UUID
    product_id: UUID | None = None
    candidates: list[VariantCandidate]
    confidence_band: str
    confidence_score: float
    missing_slots: list[str] = Field(default_factory=list)
    rationale: str | None = None


class ResolverTraceRead(BaseModel):
    id: UUID
    shop_id: UUID
    trace_type: str
    conversation_id: UUID | None = None
    input_payload: dict = Field(default_factory=dict)
    top_candidates: list[dict] = Field(default_factory=list)
    matched_aliases: list[dict] = Field(default_factory=list)
    rules_fired: list[str] = Field(default_factory=list)
    missing_slots: list[str] = Field(default_factory=list)
    confidence_band: str
    confidence_score: float
    rationale: str | None = None
    qdrant_query_metadata: dict = Field(default_factory=dict)
    created_at: datetime
    model_config = {"from_attributes": True}


class ResolverFeedbackRequest(BaseModel):
    shop_id: UUID
    action: str
    original_product_id: UUID | None = None
    corrected_product_id: UUID | None = None
    original_variant_id: UUID | None = None
    corrected_variant_id: UUID | None = None
    notes: str | None = None


class ResolverFeedbackRead(BaseModel):
    id: UUID
    shop_id: UUID
    trace_id: UUID
    action: str
    operator_id: UUID
    original_product_id: UUID | None = None
    corrected_product_id: UUID | None = None
    original_variant_id: UUID | None = None
    corrected_variant_id: UUID | None = None
    notes: str | None = None
    created_at: datetime
    model_config = {"from_attributes": True}


class ResolverConfidenceThresholds(BaseModel):
    high: float = 0.85
    medium: float = 0.55
